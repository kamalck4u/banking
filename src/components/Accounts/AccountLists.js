import React from 'react';
import { useHistory } from 'react-router-dom';
import { ACCOUNT_NUMBER, ACCOUNT_NAME, ACCOUNT_TYPE, CURRENCY, ACCOUNT_BALANCE, LOADER_TEXT_ACCOUNTS } from '../Common/Constants';
import '../../App.css';
import { Loader } from '../Common/Panels';
import { array, bool } from 'prop-types';

const AccountLists = ({ accounts, isLoading }) => {
	const history = useHistory();
	return (
		<React.Fragment>
			{
				isLoading ? <Loader loaderText={LOADER_TEXT_ACCOUNTS} />
					:
					<div className="table-responsive">
						<table className="table table-striped">
							<thead className="thead-dark">
								<tr>
									<th scope="col">{ACCOUNT_NUMBER}</th>
									<th scope="col">{ACCOUNT_NAME}</th>
									<th scope="col">{ACCOUNT_TYPE}</th>
									<th scope="col">{CURRENCY}</th>
									<th scope="col">{ACCOUNT_BALANCE}</th>
								</tr>
							</thead>
							<tbody className="row-clickable">
								{accounts.map((account) => {
									return (
										<tr key={account.id} className="clickable-row table-cell" onClick={() => history.push(`/accounts/${account.id}/transactions`, { account })}>
											<td>{account.account_number}</td>
											<td>{account.account_name}</td>
											<td>{account.account_type}</td>
											<td>{account.currency}</td>
											<td>{account.balance}</td>
										</tr>
									);
								}
								)}
							</tbody>
						</table>
					</div>
			}
		</React.Fragment>
	);
};

AccountLists.propTypes = {
	accounts: array,
	isLoading: bool
}

export default AccountLists;
