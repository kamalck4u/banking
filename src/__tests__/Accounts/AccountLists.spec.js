import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import AccountLists from '../../components/Accounts/AccountLists';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const accounts = [{
	id: '1',
	account_number: '119998888',
	account_type: 'saving plus',
	account_name: 'savings',
	currency: 'SGD',
	balance: '100'
}];

let isloading = false;
const mockHistoryPush = jest.fn();


jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

describe('AccountLists', () => {
	it('snapshot renders', () => {
		const AccountListsComponent = renderer.create(<AccountLists accounts={accounts} isLoading={isloading} />);
		expect(AccountListsComponent.toJSON()).toMatchSnapshot();
	});
});


describe('Accounts List Table', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(<AccountLists accounts={accounts} isLoading={isloading} />);
	});

	it('should show the account table headers', () => {
		expect(wrapper.find('th')).toHaveLength(5);
		expect(wrapper.find('th').getElements()[0].props.children).toEqual('Account Number');
		expect(wrapper.find('th').getElements()[1].props.children).toEqual('Account Name');
		expect(wrapper.find('th').getElements()[2].props.children).toEqual('Account Type');
		expect(wrapper.find('th').getElements()[3].props.children).toEqual('Currency');
		expect(wrapper.find('th').getElements()[4].props.children).toEqual('Balance');
	});

	it('should render the account details', () => {
		expect(wrapper.find('td')).toHaveLength(5);
		expect(wrapper.find('td').getElements()[0].props.children).toEqual('119998888');
		expect(wrapper.find('td').getElements()[1].props.children).toEqual('savings');
		expect(wrapper.find('td').getElements()[2].props.children).toEqual('saving plus');
		expect(wrapper.find('td').getElements()[3].props.children).toEqual('SGD');
		expect(wrapper.find('td').getElements()[4].props.children).toEqual('100');
	});

	it('should be clickable', () => {
		expect(wrapper.find('tbody.row-clickable')).toHaveLength(1);
		expect(wrapper.find('tr.clickable-row.table-cell')).toHaveLength(1);
	});

	it('should redirect to transaction history page', () => {
		wrapper.find('tr').at(1).simulate('click');
		expect(mockHistoryPush).toHaveBeenCalledWith('/accounts/1/transactions', {
			account: {
				id: '1',
				account_number: '119998888',
				account_type: 'saving plus',
				account_name: 'savings',
				currency: 'SGD',
				balance: '100'
			}
		});
	});

});

describe('Panel', () => {
	let wrapper;
	it('should show loading spinner while data is being fetched', () => {
		isloading = true;
		wrapper = shallow(<AccountLists accounts={accounts} isLoading={isloading} />);
		expect(wrapper.find('Loader').props()).toEqual({ loaderText: 'Your portfolio is being loaded, please wait...' });
		expect(wrapper.find('th')).toHaveLength(0);
	});
});

