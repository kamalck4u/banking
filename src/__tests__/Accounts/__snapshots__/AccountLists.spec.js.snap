// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountLists snapshot renders 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr>
        <th
          scope="col"
        >
          Account Number
        </th>
        <th
          scope="col"
        >
          Account Name
        </th>
        <th
          scope="col"
        >
          Account Type
        </th>
        <th
          scope="col"
        >
          Currency
        </th>
        <th
          scope="col"
        >
          Balance
        </th>
      </tr>
    </thead>
    <tbody
      className="row-clickable"
    >
      <tr
        className="clickable-row table-cell"
        onClick={[Function]}
      >
        <td>
          119998888
        </td>
        <td>
          savings
        </td>
        <td>
          saving plus
        </td>
        <td>
          SGD
        </td>
        <td>
          100
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
